src/util/state_lib/state_machine.py

```diff
@@ -0,0 +1,62 @@
+from .state import State, ExitState
+from typing import Dict, Set, List, Callable
+import time
+from dataclasses import dataclass
+from threading import Lock
+
+@dataclass
+class TransitionRecord:
+    time: float
+    origin_state: str
+    dest_state: str
+
+class StateMachine:
+    def __init__(self, initial_state: State, name: str):
+        self.current_state = initial_state
+        self.state_lock = Lock()
+        self.state_transitions: Dict[type[State], Set[type[State]]]  = {}
+        self.transition_log: List[TransitionRecord] = []
+        self.context = None
+        self.name = name
+    
+    def __update(self):
+        with self.state_lock:
+            current_state = self.current_state
+        next_state = current_state.on_loop(self.context)
+        if type(next_state) not in self.state_transitions[type(current_state)]:
+            raise Exception(f"Invalid transition from {current_state} to {next_state}")
+        if type(next_state) is not type(current_state):
+            current_state.on_exit(self.context)
+            self.transition_log.append(TransitionRecord(time.time(), str(current_state), str(next_state)))
```

Comment:
We can change how this works later, but I think this is not really an issue since this is actually only logging when the state changes not every state update